[Woo POS][Historical Orders] Orders Fetching

toupper opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

Part of WOOMOB-1145

Infrastructure for remotely fetching required data from API, with the implementation of a dummy UI.

For consistency, I implemented a similar approach that we use with the items. 

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

1. Open POS -> Menu -> Orders
2. Confirm the loading state appears when orders are loading
3. Confirm correct POS orders are loaded, compare with wp-admin and Woo app orders

UI is out of scope, generated to facilitate testing. In-memory cache coming next.

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

Tested on Samsung Tab A8 Android 14

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->


https://github.com/user-attachments/assets/330efadb-8bfc-4bbd-87dc-6f77f9be6b69



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

toupper committed:
Added Orders Data Source for POS

toupper committed:
Update Rest Client and Order Store to fetch POS orders

toupper committed:
Update state to fetch orders

toupper committed:
Update view model with dummy implementation for fetching orders

toupper committed:
Update UI with dummy implementation for showing fetched POS Orders

toupper committed:
Dummy layout for order rows.

toupper committed:
Dummy layout

toupper committed:
Detekt

toupper committed:
Run detekt

toupper committed:
Remove non necessary comment

toupper committed:
Add unit test

toupper committed:
Add unit tests for the store

toupper committed:
Merge branch 'trunk' into feat/WOOMOB-1145-pos-historical-orders-fetching

toupper committed:
Add unit test

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/orders/WooPosOrdersDataSource.kt:
```diff
@@ -0,0 +1,41 @@
+    private suspend fun fetchOrdersFromStore(
+        page: Int
+    ): WooResult<List<OrderEntity>> {
+        return orderStore.fetchOrders(
```
That's removes all the orders from the DB, including the orders from the main app.

I think we should probably use `OrderRestClient`

kidinov replied:
Also, shell we filter out auto drafts here?

toupper replied:
Good point! I changed it to use `OrderRestClient`. We'll filter auto drafts in a future PR

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/orders/WooPosOrdersState.kt:
```diff
@@ -1,19 +1,13 @@
-        WooPosOrder(3, "Order 3")
-    ),
-    val selectedOrderId: Long? = 1
+    val orders: List<Order> = emptyList(),
```
I mentioned [the state](https://github.com/woocommerce/woocommerce-android/pull/14522#pullrequestreview-3158630427) in the previous PR. In my opinion, the foundation of any feature is a well-defined state and it's better to start with this firstly

Each piece of state should have one clear definition, eliminating ambiguity and preventing conflicting states from existing simultaneously. Currently your state can be defined, for instance, as:

`isLoading = true` && `error = "something` && `orders = list_of_orders` which is clearly should not be possible. This is the source of many bugs

So I'd recommend to define state firstly and then go from here otherwise a lot of will be missed resulting not in well implemented feature

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/orders/WooPosOrdersScreen.kt:
```diff
@@ -55,12 +64,55 @@ fun WooPosOrdersScreen(
-                onOrderSelected = viewModel::onOrderSelected,
-                modifier = Modifier.fillMaxSize()
-            )
+            when {
```
I think this `when` is good illustration why badly defined state is a problem. You have many places here to make a mistake here just by not declaring (or declaring wrongly) one of the conditions (and no compiler check!) so you can end up in "else" case when it's not needed.

Also, `state.error != null` and `state.orders.isEmpty()` is kinda not view level logic - should be in the VM

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/orders/WooPosOrdersScreen.kt:
```diff
@@ -55,12 +64,55 @@ fun WooPosOrdersScreen(
-            )
+            when {
+                state.isLoading -> {
+                    Column(
```
I recommend extracting all states here into separate composable functions with proper naming for better readability and code maintainability

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/orders/WooPosOrdersDataSource.kt:
```diff
@@ -0,0 +1,41 @@
+
+    private suspend fun fetchOrdersFromStore(
+        page: Int
+    ): WooResult<List<OrderEntity>> {
```
I think it'd better to return either manually defined result or `kotlin.Result`. Ideally your own sealed class result, as `kotlin.Result` has it's own design issues as it is implemented as an inline value class wrapping Any?, but conceptually behaves like a sealed class with Success and Failure cases. This creates friction in the type system similar to `WooResult`.

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/orders/WooPosOrdersScreen.kt:
```diff
@@ -40,6 +45,10 @@ fun WooPosOrdersScreen(
     val viewModel: WooPosOrdersViewModel = hiltViewModel()
     val state by viewModel.state.collectAsState()
 
+    LaunchedEffect(Unit) {
```
That will refetch all the orders on any configuration change. Any reason to not fetch them in VM's `init`?

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/orders/WooPosOrdersScreen.kt:
```diff
@@ -86,29 +138,51 @@ fun WooPosOrdersListPaneScreen(
+                        style = WooPosTypography.BodyMedium
+                    )
+                    WooPosText(
+                        text = order.dateCreated.formatToDDMMMYYYY(),
```
Should not be part of the view. All the data that comes to the view should be already ready to be displayed

kidinov reviewed (commented):
@toupper Overall, it looks great! Thanks! I have left a few suggestions. But firstly, I would suggest taking a step back and redesigning the state before proceeding further. This will save a lot of rework eventually

toupper commented:
Thanks for the review and comments @kidinov! They are very instructive and I will use them in future PRs.

As noted in the description, the UI (including state and view model) isn’t meant to be part of this PR. The UI code was added quickly just to help test fetching, not meant to be production-ready.

That said, I understand if you’d prefer not to merge code that feels suboptimal. One option would be to close this PR and open a new one focused only on the fetching logic, with the bare minimum needed to verify it works (e.g., logging the results).

Would you prefer that approach?

kidinov commented:
@toupper, I’m completely fine with merging it if it works for you this way. However, there are comments not only about the state but also about fetching and modifying the persistence of the main app, which we probably shouldn’t do

toupper committed:
Use rest client instead of the WCOrderStore, to avoid modifying the DB

toupper committed:
Revert changes in the WCOrderStore

toupper committed:
Use our own sealed class for the return type

toupper committed:
Use LoadOrdersResult instead of WooResult, which has design issues

toupper committed:
Move refreshOrders call to view model init, to avoid it being called on every configuration change.

toupper commented:
@kidinov Could you please review it again? I’ve addressed all the comments except those specifically related to UI and state, which I plan to handle in the next PRs.

toupper committed:
Better naming

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/orders/WooPosOrdersDataSource.kt:
```diff
@@ -0,0 +1,63 @@
+
+        return if (result.isError) {
+            FetchOrdersResult.Error(
+                WooError(API_ERROR, SERVER_ERROR, result.error.message)
```
How do you know that it's API error and not lets say network? I think if you don't need to expose specific error it's fine but at least we should not return incorrect/misleading error

Also, not sure if we actually need to have 2 return types for this, do we? Maybe something simple like that:

```
sealed class LoadOrdersResult {
    data class Success(val orders: List<Order>) : LoadOrdersResult()
    data class Error(val message: String) : LoadOrdersResult()
}

class WooPosOrdersDataSource @Inject constructor(
    private val restClient: OrderRestClient,
    private val selectedSite: SelectedSite,
    private val orderMapper: OrderMapper,
) {
    suspend fun loadOrders(): LoadOrdersResult {
        val result = restClient.fetchOrders(
            site = selectedSite.get(),
            count = 25,
            page = 1,
            orderBy = OrderBy.DATE,
            sortOrder = OrderRestClient.SortOrder.DESCENDING,
            statusFilter = null,
            createdVia = "pos-rest-api"
        )

        return if (result.isError) {
            LoadOrdersResult.Error(result.error.message)
        } else {
            LoadOrdersResult.Success(result.orders.toAppModels())
        }
    }

    private suspend fun List<OrderEntity>.toAppModels(): List<Order> = map { 
        orderMapper.toAppModel(it) 
    }
}
```

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/orders/WooPosOrdersState.kt:
```diff
@@ -1,19 +1,13 @@
+    val error: String? = null
 ) {
-    val selectedOrder: WooPosOrder?
+    val selectedOrder: Order?
```
Not for this PR, so just to make sure that it won't be forgotten - we should have our `WooPosOrder` that contains info only needed to be shown. The order is data layer model

kidinov reviewed (approved):
LGTM! I left a couple of minor suggestions, please take a look!

toupper committed:
better naming

toupper committed:
Fix lint issue

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-android/milestone/267">23.2</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

toupper committed:
Refactor error handling

toupper committed:
Fix test
